test/cpp/util/channelz_sampler_test.cc

```diff
@@ -0,0 +1,172 @@
+  grpc::ServerBuilder builder;
+  auto server_creds =
+      grpc::testing::GetCredentialsProvider()->GetServerCredentials(
+          custom_credentials_type);
+  builder.AddListeningPort(server_address, server_creds);
+
+  // forces channelz and channel tracing to be enabled.
+  builder.AddChannelArgument(GRPC_ARG_ENABLE_CHANNELZ, 1);
+  builder.AddChannelArgument(GRPC_ARG_MAX_CHANNEL_TRACE_EVENT_MEMORY_PER_NODE,
+                             1024);
+  builder.RegisterService(&service);
+  std::unique_ptr<Server> server(builder.BuildAndStart());
+  gpr_log(GPR_INFO, "Server listening on %s", server_address.c_str());
+  server->Wait();
+}
+
+// Creata an echo client - set timeout as 0.15s
+class EchoClientImpl {
+ public:
+  EchoClientImpl(std::shared_ptr<Channel> channel)
+      : stub_(grpc::testing::TestService::NewStub(channel)) {}
+  Status EmptyCall() {
+    grpc::testing::Empty request;
+    grpc::testing::Empty response;
+    ClientContext context;
+    int64_t timeout_microseconds = 150;
+    context.set_deadline(
+        grpc_timeout_milliseconds_to_deadline(timeout_microseconds));
+    Status status = stub_->EmptyCall(&context, request, &response);
+    return status;
+  }
+
+ private:
+  std::unique_ptr<grpc::testing::TestService::Stub> stub_;
+};
+
+// Run client in a thread
+void RunClient(std::string client_id) {
+  // std::string target_str = "localhost:10000";
+  grpc::ChannelArguments channel_args;
+  std::shared_ptr<grpc::ChannelCredentials> channel_creds =
+      grpc::testing::GetCredentialsProvider()->GetChannelCredentials(
+          custom_credentials_type, &channel_args);
+  EchoClientImpl echoer(grpc::CreateChannel(server_address, channel_creds));
+  unsigned int client_echo_sleep_second = 1;
+
+  gpr_log(GPR_INFO, "Client %s is echoing!", client_id.c_str());
+  while (true) {
+    Status status = echoer.EmptyCall();
+    sleep(client_echo_sleep_second);
+  }
+}
+
+// Test the channelz sampler
+TEST(ChannelzSamplerTest, SimpleTest) {
+  // server thread
+  std::thread server_thread(RunServer);
+  gpr_log(GPR_INFO, "Wait 3 seconds to make sure server is started...");
+  float server_start_seconds = 3.0;
+  sleep(server_start_seconds);
+
+  // client thread
+  std::thread client_thread_1(RunClient, "1");
+  std::thread client_thread_2(RunClient, "2");
+
+  // Run the channelz sampler
+  std::string channelz_sampler_bin_path =
+      "./bazel-bin/test/cpp/util/channelz_sampler";
+  grpc::SubProcess* test_driver = new grpc::SubProcess(
+      {std::move(channelz_sampler_bin_path),
+       "--server_address=" + server_address,
+       "--custom_credentials_type=" + custom_credentials_type,
+       "--sampling_times=" + sampling_times,
+       "--sampling_interval_seconds=" + sampling_interval_seconds,
+       "--output_json=" + output_json});
+
+  test_driver->Join();
+  server_thread.join();
+  client_thread_1.join();
+  client_thread_2.join();
+
+  EXPECT_TRUE(true);
```

Comment:
this is a no-op, right? can we remove this?

Reply:
Removed